src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java

```diff
@@ -19,4 +20,7 @@
     @EntityGraph(attributePaths = { "submissions", "results.submission" })
     @Query("select pe from Participation pe")
     List<Participation> getAllWithEagerSubmissionsAndResults();
+
+    @Query("select distinct participation from Participation participation left join fetch participation.submissions submission left join fetch participation.results where participation.student.login = :#{#username}")
+    List<StudentParticipation> findByStudentUsernameWithEagerSubmissionsWithEagerResults(@Param("username") String username);
```

Comment:
I think this db call is too expensive. Therefore, I would suggest that we only load the participations together with their latest submission (simply take the submission date into account) left join submission.result (which might be null).

Reply:
I just thought about it and it might be wrong to take the last submission, so I suggest we take all submissions together with their result. This would be 

`    @Query("select distinct p from StudentParticipation p left join fetch p.submissions s left join fetch s.result r where p.student.id = :#{#studentId}")
    List<StudentParticipation> findAllByStudentIdWithSubmissionsWithResult(@Param("studentId") Long studentId);`

We can later on think about if we can already filter on the database level which submission should be used or if we want to store this in the database.

Reply:
I have added a new call into the corresponding method. It will lead to some changes because the participation then does not contain results directly, but submissions which might include a result.

I think we then need to find the "correct" submission + result to display. We need to take the following into account (which might not be fully complete):

1) If there is at least one submission with a rated result, we filter for submissions with rated results and take the submission with the latest submission date (independent of the exercise date)
--> this means the result shown in the client is rated/graded
(It might be the case that we have later submissions which are unrated/ungraded, but we ignore them)

2) If there are results, but there is no submission with a rated result, we take the submission with the latest submission date. 
--> this means the submission is unrated/ungraded, because the submission was late.

3) If there are submissions, but no results yet, we take the submission with the latest submission date
--> this means the submission is not graded yet (in the future we might show the submission status / submission date then)
(Note: this case might be tricky in the future, because one submission could be before the due date and would be rated in the future, and another submission could be after the due date and would be unrated in the future. While this case might not occur yet, it could theoretically occur in the future)

Reply:
In general, I wonder if it won't lead to even more changes if we (a) don't send all results anymore and (b) package the *right* result in a submission.

In the client `this.courseService.findAll()`, that uses the `/for-dashboard/` route, is used more than a dozen times. I already tried today to see if the client somewhere really needs all the results, but I'm still not sure. @krusche You have the best overview... do you think that would be a problem?

For (b) it might make sense to place the *right* result into the `Participation.results`.